docs/sql-ref-datatypes.md

```diff
@@ -314,3 +314,51 @@ SELECT COUNT(*), c2 FROM test GROUP BY c2;
 |        3| Infinity|
 +---------+---------+
 ```
+
+#### Data type compatibility
```

Comment:
`Data type compatibility` => `Type Conversion`?

Reply:
How about organizing the structure of this section by referring to the Oracle doc like this? https://docs.oracle.com/cd/B28359_01/server.111/b28286/sql_elements002.htm#SQLRF51043
```
#### Type Conversion
<What does "Type Conversion" means? How does Spark handle type conversion? brabrabra...>

#### Type Coercion
<Coercion Matrix> 

##### Type Coercion Examples
<examples...>

#### Explicit Casting and Store Assignment Casting
<Casting Matrix> 

##### Type Casting Examples
<examples...>
```
cc: @gatorsmile @HyukjinKwon @huaxingao